src/zenml/orchestrators/step_runner.py

```diff
@@ -647,6 +657,91 @@ def _link_artifacts_to_model(
+            ModelVersionPipelineRunRequestModel,
+        )
+
+        zs = Client().zen_store
+
+        for artifact_name in artifact_names:
+            artifact_config = (
+                get_step_context()._get_output(artifact_name).artifact_config
+            )
+            if artifact_config is not None:
+                try:
+                    model_id = (
+                        artifact_config._model_config.get_or_create_model().id
+                    )
+                    model_version_id = (
+                        artifact_config._model_config._get_model_version().id
+                    )
+                except RuntimeError:
+                    break
+                else:
+                    zs.create_model_version_pipeline_run_link(
+                        ModelVersionPipelineRunRequestModel(
+                            user=Client().active_user.id,
+                            workspace=Client().active_workspace.id,
+                            name=pipeline_run.name,
+                            pipeline_run=pipeline_run.id,
+                            model=model_id,
+                            model_version=model_version_id,
+                        )
+                    )
+        for external_artifact in external_artifacts:
+            if (
+                external_artifact.model_artifact_name is not None
+                and external_artifact.model_name is not None
+            ):
+                if external_artifact.model_version is not None:
+                    model_version = zs.get_model_version(
+                        model_name_or_id=external_artifact.model_name,
+                        model_version_name_or_id=external_artifact.model_version,
+                    )
+                else:
+                    model_version = zs.get_model_version(
+                        model_name_or_id=external_artifact.model_name
+                    )
+                zs.create_model_version_pipeline_run_link(
+                    ModelVersionPipelineRunRequestModel(
+                        user=Client().active_user.id,
+                        workspace=Client().active_workspace.id,
+                        name=pipeline_run.name,
+                        pipeline_run=pipeline_run.id,
+                        model=model_version.model.id,
+                        model_version=model_version.id,
+                    )
+                )
+
+        try:
+            mc = get_step_context().model_config
+        except StepContextError:
+            return
+
+        zs.create_model_version_pipeline_run_link(
+            ModelVersionPipelineRunRequestModel(
+                user=Client().active_user.id,
+                workspace=Client().active_workspace.id,
+                name=pipeline_run.name,
+                pipeline_run=pipeline_run.id,
+                model=mc.get_or_create_model().id,
+                model_version=mc._get_model_version().id,
+            )
+        )
```

Comment:
Maybe this could be made a bit more readable:

```suggestion
    def _link_pipeline_run_to_model(
        self,
        pipeline_run: "PipelineRunResponseModel",
        artifact_names: List[str],
        external_artifacts: List["ExternalArtifactConfiguration"],
    ) -> None:
        """Links the pipeline run to the model version.

        Args:
            pipeline_run: The response model of current pipeline run.
            artifact_names: The name of the published output artifacts.
            external_artifacts: The external artifacts of the step.
        """
        from zenml.models.model_models import (
            ModelVersionPipelineRunRequestModel,
        )
        
        models: List[Tuple[UUID, UUID]] = []  # (model_id, model_version_id)

        zs = Client().zen_store

        for artifact_name in artifact_names:
            artifact_config = (
                get_step_context()._get_output(artifact_name).artifact_config
            )
            if artifact_config is not None:
                try:
                    model_id = (
                        artifact_config._model_config.get_or_create_model().id
                    )
                    model_version_id = (
                        artifact_config._model_config._get_model_version().id
                    )
                    models.append((model_id, model_version_id))
                except RuntimeError:
                    break

        for external_artifact in external_artifacts:
            if (
                external_artifact.model_artifact_name is not None
                and external_artifact.model_name is not None
            ):
                if external_artifact.model_version is not None:
                    model_version = zs.get_model_version(
                        model_name_or_id=external_artifact.model_name,
                        model_version_name_or_id=external_artifact.model_version,
                    )
                else:
                    model_version = zs.get_model_version(
                        model_name_or_id=external_artifact.model_name
                    )
                models.append(model_version.model_id, model_version.id)

        try:
            mc = get_step_context().model_config
            models.append((mc.get_or_create_model().id, mc._get_model_version().id))
        except StepContextError:
            pass
        
        for model_id, model_version_id in models:
            zs.create_model_version_pipeline_run_link(
                ModelVersionPipelineRunRequestModel(
                    user=Client().active_user.id,
                    workspace=Client().active_workspace.id,
                    name=pipeline_run.name,
                    pipeline_run=pipeline_run.id,
                    model=model_id,
                    model_version=model_version_id
            )
        )
```

Reply:
Or maybe also nice to split it up into several smaller functions, e.g., `_link_pipeline_run_to_models_of_artifacts`, `_link_pipeline_run_to_models_of_external_artifacts`?

Reply:
Refactored, following both suggestions